This commit enables the following features:

ovanov opened the pull request:
- Test suite for UnitTests which cover the stream topology
- enhanced docker-compose.yaml: now it does not run the application by default
- Refactoring of Stream Topology into its own class, called DefaultStreamsTopology. The class is also a @Component and is also included in the application context and is therefore used by the AuthorizationDeduplicationStream.buildPipeline method
- updated kafka-streams-test and junit dependencies in pom file
- Created new helper class StreamsUtils as Confluent recommends it for avro based value tests in kafka streams

Commit:
This commit enables the following features:
- Test suite for UnitTests which cover the stream topology
- enhanced docker-compose.yaml: now it does not run the application by default
- Refactoring of Stream Topology into its own class, called DefaultStreamsTopology. The class is also a @Component and is also included in the application context and is therefore used by the AuthorizationDeduplicationStream.buildPipeline method
- updated kafka-streams-test and junit dependencies in pom file
- Created new helper class StreamsUtils as Confluent recommends it for avro based value tests in kafka streams

mirkoipt commented on src/test/java/com/ipt/kafkatopicupdates/streams/DefaultStreamsTopologyTest.java:
```diff
@@ -0,0 +1,270 @@
+        config.put(StreamsConfig.APPLICATION_ID_CONFIG, "test");
+
+        config.put("schema.registry.url", "mock://test");
+
```
Why not extract the Properties' creation to the Util class?

ovanov replied:
Good idea

mirkoipt commented on src/test/java/com/ipt/kafkatopicupdates/streams/DefaultStreamsTopologyTest.java:
```diff
@@ -0,0 +1,270 @@
+        serdeConfig = StreamsUtils.propertiesToMap(config);
+        authorizationSerde = StreamsUtils.getSpecificAvroSerde(serdeConfig);
+
+        config.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.String().getClass());
```
Why is this config different from the config the serde has? Couldn't the serdeConfig also contain this additional info? If possible, I would use the same config.

ovanov replied:
The serde config is a `Map<String, Object>`, the `config` variable, on the other hand, is a `Properties()`object. To create a `SpecificAvroSerde` we need the Map. To create a `TopologyTestDriver`we require a Property object.

mirkoipt commented on src/main/java/com/ipt/kafkatopicupdates/streams/DefaultStreamsTopology.java:
```diff
@@ -0,0 +1,49 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class DefaultStreamsTopology {
```
IMO this is not a "default" Topology. It is a very specific Topology. e.g. AuthorizationDeduplicationTopology

ovanov replied:
True, will change it

mirkoipt commented on src/main/java/com/ipt/kafkatopicupdates/streams/DefaultStreamsTopology.java:
```diff
@@ -0,0 +1,49 @@
+
+import org.springframework.stereotype.Component;
+
+@Component
```
Why is this a component? Isn't this "de facto" another Utility class?

ovanov replied:
Actually yes. I chose the `@Component` annotation to include the class in the Application Context. This enables simple DI in the `AuthenticationDeduplicationStreams.java` class. It does not need to know about the Topology. Testing is also simpler, if we want to use our own topology.

mirkoipt commented on src/main/java/com/ipt/kafkatopicupdates/producer/AuthorizationAvroProducer.java:
```diff
@@ -36,7 +36,7 @@ private void scheduleFixedRateTask() {
 
         ListenableFuture<SendResult<String, Authorization>> future =
-                authorizationKafkaTemplate.send(
+                (ListenableFuture<SendResult<String, Authorization>>) authorizationKafkaTemplate.send(
```
Does this throw a warning in the IDE? My IDE says it's unnecessary.

ovanov replied:
Good point. I think in an earlier version I needed to set this. Now it is not needed anymore.

mirkoipt commented on src/test/java/com/ipt/kafkatopicupdates/streams/DefaultStreamsTopologyTest.java:
```diff
@@ -0,0 +1,270 @@
+    private DefaultStreamsTopology defaultStreamsTopology;
+
+    private Properties config;
+    private Map<String, Object> serdeConfig;
```
config and serdeConfig are static and therefore do not have to be initialized in @BeforeEach. It can be done here. This only applies if they stay the same for the tests, though.

ovanov replied:
True!

mirkoipt commented on src/test/java/com/ipt/kafkatopicupdates/streams/DefaultStreamsTopologyTest.java:
```diff
@@ -0,0 +1,270 @@
+    private TestOutputTopic<String, Authorization> outputTopic;
+
+    private final Serde<String> stringSerde = Serdes.String();
+    private SpecificAvroSerde<Authorization> authorizationSerde;
```
authorizationSerde can be static and can therefore be defined here.

ovanov replied:
True, this is a good point.

Commit:
Changes:
- Deleted unused Application Test Class
- Refactored attributes out of the `setUp()` method, due to them being final and no state change is happening
- Annotated Test Cases with a displayed name, in order to give a documenting purpose
- Renamed DefaultStreamsTopology.java to AuthorizationDeduplicationTopology.java. Reason being that the topology is no boilerplate, but a concrete example
- Changed StreamsUtils to no provide a `serdesConfig` of type `Properties` while keeping the `StreamsUtils.getSpecificAvroSerde` method
- Added a private constructor to StreamsUtils. Utility classes should not have public constructors. I used the one implementation that SonarLint suggests
- changed new line breaks in parameters of AuthorizationDeduplicationStream.buildPipeline for readability reasons.
